lean4/app/Main.lean

```diff
@@ -0,0 +1,653 @@
+  (blockCallbacks  : Fn.Name -> Block.Label -> (ProverInterface -> IO Unit) -> IO Unit)
+  (sessionComplete : IO Unit)
+
+
+structure ModuleVCG.Context :=
+  (moduleAnn : Module.Annotation)
+  -- ^ Annotations for module.
+  (moduleMem : Memory 64)
+  -- ^ Machine code memory
+  (symbolAddrMap : RBMap ByteString (MemSegmentOff 64) (λ x y => x < y))
+  -- ^ Maps bytes to the symbol name
+  (writeStderr : Bool)
+  -- ^ Controls whether logs, warnings or errors
+  -- chould be written to stderr.
+  --
+  -- If false, the messages are droped, but warning
+  -- count is increased.
+  (errorCount : IO.Ref Nat)
+  -- ^ Counts numbers of warnings generated during
+  -- verification for display at end of run.
+  (proverGen : ProverSessionGenerator)
+  -- ^ Interface for generating prover sessions.
+  (moduleTypeMap : (RBMap LLVM.Ident LLVM.Ty (λ x y => x < y)))
+  -- ^ type map for module.
+
+
+
+namespace Block
+
+inductive Error
+| AnnParseFailure : String → Error
+| MissingAnnotations : Error
+| UnsupportedPhiVarType : Ident → Ty → Error
+| BlockAddrInvalid : MemWord 64 → Error
+
+namespace Error
+
+def toString : Error → String
+| AnnParseFailure msg => "Annotation parse failure: " ++ msg
+| MissingAnnotations => "Could not find block annotations."
+| UnsupportedPhiVarType x t => "Phi variable "++x.toString++" has unsupported type "++t.toString++"."
+| BlockAddrInvalid addr => "Annotated block address "++addr.toString++" is not not in code segment."
+
+end Error
+
+end Block
+
+
+
+
+namespace Module
+
+inductive Error
+| Custom : String → Error
+| Function : Fn.Name → Fn.Error → Error
+| Block : Fn.Name → Block.Label → Block.Error → Error
+| IO : IO.Error -> Error
+
+namespace Error
+
+def toString : Error → String
+| Custom msg => msg
+| Function fnm ferr => fnm++". "++ferr.toString
+| Block fnm lbl err => fnm++"."++lbl.toString++". "++err.toString
+| IO err => err.toString
+
+def toIOError : Error → IO.Error
+| Error.IO ioErr => ioErr
+| e => IO.userError $ "Uncaught module VCG error: " ++ e.toString
+
+
+def liftIO {α} (m : EIO IO.Error α) : EIO Error α := 
+  m.adaptExcept Error.IO
+
+
+instance : MonadIO (EIO Error) := {monadLift := @liftIO}
+
+
+end Error
+
+end Module
+
+
+-- Pretty print an error that occurs at the start of an instruction.
+def renderMCInstError
+ (fnm  : Fn.Name) -- ^ Name of function
+ (lbl  : Block.Label) -- ^ Block label
+ (idx  : Int) -- ^ LLVM instruction index
+ (addr : MemSegmentOff 64) -- ^ Address of current instruction.
+ (msg  : String) -- ^ Message
+ : String :=
+  fnm++"."++lbl.toString++"."++idx.repr++" ("++addr.toString++"). "++msg
+  -- ^^ FIXME `addr.toString` here was `showsPrec 10 addr ""`... does that matter?
+
+
+
+
+-------------------------------------------------------
+-- ModuleVCG
+-------------------------------------------------------
+
+
+-- A monad for running verification of an entire module
+@[reducible]
+def ModuleVCG := ReaderT ModuleVCG.Context (EIO Module.Error)
```

Comment:
After some more thought, it would be ideal from my perspective to move IO and EIO out of this monad, and replace it with a pure monad that will yield all the events in a lazy way.  For example, rather than the IORef and ProofSession generator, we yield potentially structured events that let a client running the ModuleVCG to respond in some way.

The motivation is that per [this discussion](https://leanprover.zulipchat.com/#narrow/stream/147302-lean4/topic/io.20monad.20mess), `IO` will have very little that is provable about it.
Is this enough guidance?  I think the change would be to drop the `IO` references from `ModuleVCG.Context`, introduce a `ModuleVCG.Event`, and make `ModuleVCG` as something like `ModuleVCG.Context -> ModuleVCG.Event`.  `ModuleVCG.Event` constructors may have continuations `Context -> Event` embedded in them.  There are likel termination issues with this, so we could see about lazy lists on the Lean4 channel.